pkg/agent/core/state.go

```diff
@@ -429,7 +457,21 @@ func (s *state) calculateNeonVMAction(
 	desiredResources api.Resources,
 	pluginRequested *api.Resources,
 	pluginRequestedPhase string,
+	logicalTime *vmv1.LogicalTime,
 ) (*ActionNeonVMRequest, *time.Duration) {
+
+	desiredTime := logicalTime
+
+	if desiredResources.HasFieldLessThan(s.VM.Using()) {
+		// We are downscaling, so we needed a permit from the¡ monitor
+		desiredTime = desiredTime.Earliest(s.Monitor.CurrentLogicalTime)
+	}
+
+	if desiredResources.HasFieldGreaterThan(s.VM.Using()) {
+		// We are upscaling, so we needed a permit from the plugin
+		desiredTime = desiredTime.Earliest(s.Plugin.CurrentLogicalTime)
+	}
```

Comment:
I think this is not sound under the following sequence of events:

1. At some prior time, we scaled down with logical time `A`.
1. We decide to scale up, recording logical time `B`
1. Record logical time `B`
1. Scheduler plugin approves scale-up with logical time `B`
1. We update neonvm to logical time `B` for the scale-up
1. We change our minds, deciding actually to scale back down. New logical time `C`
1. We update neonvm, but because it's scaling down, we use `min(A, C)` because we never got around to updating the vm-monitor

So, we end up with an arbitrarily old desired time if we change our minds (?)

Not sure if I missed something, or if this is actually ok — lmk!

Reply:
So, in this case:
 - After step 1 we have current=desired=A, so we've measured the latency for normal scale down.
 - After step 5 we have status.current=A and spec.desired=B.
    - If neonvm-controller will observe this, it will propagate status.current=B, this will loop back to the agent, and we observe latency for normal scale down.
    - If neonvm-controller doesn't observe this, nothing changes.
 - After step 7
    - If current time was propagated to B, then we have a weird situation where the desired time is less than current. This should be fixed (see note below).
    - If current time wasn't propagated to B between steps 5 and 6, then desired=A and current=A, and also we don't need to do any work (wanted to upscale, but decided not to).
 
Regardless, no new measurement will be added, since we already measured the latency to get current=A, and we can't measure the same latency again. This is equivalent to saying "steps 2-7 are actually part of the scale down, which happened in step 1, because they are started by that monitor request".

But what do we even want to measure in this case? I think measuring only clean scale down during step 1 (and, potenitally, scale-up 2-4, if it completed) is fine, otherwise it becomes harder to reason about what are we even measuring.

Note how to fix reversed time flow:
 - The interpretation is "we want neonvm to reach the state which was intended before neonvm got to it's current state".
 - I think, just check `desired > current` before doing `current = desired` should be fine.

I don't know if what I wrote makes sense to you, feel free to ask more.

Reply:
> I think measuring only clean scale down during step 1 (and, potenitally, scale-up 2-4, if it completed) is fine, otherwise it becomes harder to reason about what are we even measuring.

Yep, +1

> I think, just check `desired > current` before doing `current = desired` should be fine.

Not sure which you're referring to, but if you think it fixes the issue, good by me :)